src/ir/eh-utils.cpp

```diff
@@ -16,53 +16,137 @@
+      // in binary, it is fine if a 'pop' is in there. We don't allow a 'pop' to
+      // be in an 'if''s then or else body because they are not
+      // first descendants.
+      if (auto* if_ = firstChild->dynCast<If>()) {
+        firstChild = if_->condition;
+        continue;
+      }
+      // We don't allow the pop to be included in a loop, because it cannot be
+      // run more than once
+      if (firstChild->is<Loop>()) {
+        return nullptr;
+      }
+      if (firstChild->is<Block>()) {
+        // If there are no branches that targets the implicit block, it will be
+        // removed when written back. But there are branches that target the
+        // implicit block,
+        // (catch $e
+        //   (block $l0
+        //     (pop i32) ;; within a block!
+        //     (br $l0)
+        //     ...
+        //   )
+        // This cannot be removed, so this is considered a nested pop (which we
+        // should fix).
+        if (firstChild == implicitBlock) {
+          if (BranchUtils::BranchSeeker::has(implicitBlock,
+                                             implicitBlock->name)) {
+            isPopNested = true;
+          }
+        } else {
+          isPopNested = true;
+        }
+      }
+      if (firstChild->is<Try>()) {
+        isPopNested = true;
+      }
+    }
+    ChildIterator it(firstChild);
     if (it.begin() == it.end()) {
-      return false;
+      return nullptr;
+    }
+    firstChildPtr = &*it.begin();
+    firstChild = *firstChildPtr;
+  }
+}
+
+bool isPopValid(Expression* catchBody) {
+  bool isPopNested = false;
+  Expression** popPtr = nullptr;
+  auto* pop = getFirstPop(catchBody, isPopNested, popPtr);
+  return pop != nullptr && !isPopNested;
+}
+
+void handleBlockNestedPops(Function* func, Module& wasm) {
+  Builder builder(wasm);
+  FindAll<Try> trys(func->body);
+  for (auto* try_ : trys.list) {
+    for (Index i = 0; i < try_->catchTags.size(); i++) {
+      Name tagName = try_->catchTags[i];
+      auto* tag = wasm.getTag(tagName);
+      if (tag->sig.params == Type::none) {
+        continue;
+      }
+
+      auto* catchBody = try_->catchBodies[i];
+      bool isPopNested = false;
+      Expression** popPtr = nullptr;
+      Expression* pop = getFirstPop(catchBody, isPopNested, popPtr);
+      assert(pop && "Pop has not been found in this catch");
+
+      // Change code like
+      // (catch $e
+      //   ...
+      //   (block
+      //     (pop i32)
+      //   )
+      // )
+      // into
+      // (catch $e
+      //   (local.set $new
+      //     (pop i32)
+      //   )
+      //   ...
+      //   (block
+      //     (local.get $new)
+      //   )
+      // )
+      if (isPopNested) {
+        assert(popPtr);
+        Index newLocal = builder.addVar(func, pop->type);
+        try_->catchBodies[i] =
+          builder.makeSequence(builder.makeLocalSet(newLocal, pop), catchBody);
+        *popPtr = builder.makeLocalGet(newLocal, pop->type);
+      }
```

Comment:
This should call `handleNondefaultableLocals` as well for the non-nullable local case, as I assume it's possible to pop a non-nullable type?

(All these corner cases stack up, sadly...)

Reply:
Thanks. I added `$pop-non-defaultable-type-within-block` to test this case too.